akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/PersistentBehaviors6.scala

```diff
@@ -21,6 +20,118 @@ API experiment with factory for command and event handler
 
 object PersistentBehaviors6 {
 
+  class HandlerFactory[Command, Event, State] {
+    type Effect = akka.persistence.typed.scaladsl.Effect[Event, State]
+
+    type CommandHandler = Command ⇒ Effect
+    type EventHandler = Event ⇒ State
+
+    object CommandHandler {
+
+      def apply(handler: Command ⇒ Effect): Command ⇒ Effect = handler
+
+      def apply(handler: (State, Command) ⇒ Effect): (State, Command) ⇒ Effect = handler
+
+      def unhandled: Command ⇒ Effect = _ ⇒ Effect.unhandled
+
+      def partial(handler: PartialFunction[Command, Effect]): Command ⇒ Effect = {
+        handler.orElse[Command, Effect] {
+          case _ ⇒ Effect.unhandled
+        }
+      }
+
+      def partial(handler: PartialFunction[(State, Command), Effect]): (State, Command) ⇒ Effect = {
+        val totalHandler = handler.orElse[(State, Command), Effect] {
+          case _ ⇒ Effect.unhandled
+        }
+        (state, command) ⇒ totalHandler((state, command))
+      }
+    }
+
+    object EventHandler {
+
+      def apply(handler: Event ⇒ State): Event ⇒ State = handler
+
+      def apply(handler: (State, Event) ⇒ State): (State, Event) ⇒ State = handler
+
+      def unhandled: Event ⇒ State = event ⇒ throw new IllegalStateException(s"unexpected event [${event.getClass}]")
+
+      def partial(handler: PartialFunction[Event, State]): Event ⇒ State =
+        handler.orElse {
+          case event ⇒
+            throw new IllegalStateException(s"unexpected event [${event.getClass}]")
+        }
+
+      def partial(handler: PartialFunction[(State, Event), State]): (State, Event) ⇒ State = {
+        val totalHandler = handler.orElse[(State, Event), State] {
+          case event ⇒
+            throw new IllegalStateException(s"unexpected event [${event.getClass}]")
+        }
+
+        (state, event) ⇒ totalHandler((state, event))
+      }
+    }
+
+  }
+
+  /**
+   * This HandlerFactory is designed to be used with PersistentBehaviors define
+   * with a Option state, ie: Option[MyModel] instead of just MyModel.
+   *
+   */
+  class HandlerFactoryOption[Command, Event, State] {
```

Comment:
This is a HandlerFactory optimise for Optional models. 

User encodes the handlers using the raw model, but the factory lift then to Option so it aligns with the Behaviour

Reply:
I'm pleased to see this solution for removing the Option boilerplate. It's great that the core API is minimalistic and these are only helpers. i think I prefer some of the more explicit alternatives, also for `Option[State]` but I wouldn't mind having these kind of helpers.